File: lib/presentation/widgets/timeline.dart
import 'package:flutter/material.dart';

class TimelineEntry {
  const TimelineEntry({
    required this.title,
    required this.subtitle,
    this.done = false,
  });

  final String title;
  final String subtitle;
  final bool done;
}

/// A vertical timeline / activity feed with connected nodes.
class Timeline extends StatelessWidget {
  const Timeline({super.key, required this.entries});

  final List<TimelineEntry> entries;

  @override
  Widget build(BuildContext context) {
    final scheme = Theme.of(context).colorScheme;
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisSize: MainAxisSize.min,
      children: [
        for (var i = 0; i < entries.length; i++)
          IntrinsicHeight(
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Column(
                  children: [
                    Container(
                      width: 18,
                      height: 18,
                      decoration: BoxDecoration(
                        color: entries[i].done
                            ? scheme.primary
                            : scheme.surfaceContainerHighest,
                        shape: BoxShape.circle,
                        border: Border.all(color: scheme.primary, width: 2),
                      ),
                      child: entries[i].done
                          ? const Icon(Icons.check,
                              size: 11, color: Colors.white)
                          : null,
                    ),
                    if (i != entries.length - 1)
                      Expanded(
                        child: Container(
                          width: 2,
                          color: scheme.outlineVariant,
                        ),
                      ),
                  ],
                ),
                const SizedBox(width: 14),
                Expanded(
                  child: Padding(
                    padding: const EdgeInsets.only(bottom: 22),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(
                          entries[i].title,
                          style: const TextStyle(
                              fontWeight: FontWeight.w600, fontSize: 15),
                        ),
                        const SizedBox(height: 2),
                        Text(
                          entries[i].subtitle,
                          style: TextStyle(
                              color: scheme.onSurfaceVariant, fontSize: 13),
                        ),
                      ],
                    ),
                  ),
                ),
              ],
            ),
          ),
      ],
    );
  }
}

File: android_native/app/src/main/java/com/example/widgetgallery/presentation/components/Timeline.kt
package com.example.widgetgallery.presentation.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

data class TimelineEntry(val title: String, val subtitle: String,
    val done: Boolean = false)

/** Vertical timeline / activity feed with connected nodes. */
@Composable
fun Timeline(entries: List<TimelineEntry>, modifier: Modifier = Modifier) {
    val scheme = MaterialTheme.colorScheme
    Column(modifier) {
        entries.forEachIndexed { i, e ->
            Row {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Box(
                        Modifier.size(18.dp).clip(CircleShape)
                            .background(
                                if (e.done) scheme.primary
                                else scheme.surfaceContainerHighest
                            )
                            .border(2.dp, scheme.primary, CircleShape),
                        contentAlignment = Alignment.Center,
                    ) {
                        if (e.done) Icon(Icons.Filled.Check, null,
                            tint = Color.White, modifier = Modifier.size(11.dp))
                    }
                    if (i != entries.lastIndex) {
                        Box(Modifier.width(2.dp).height(46.dp)
                            .background(scheme.outlineVariant))
                    }
                }
                Spacer(Modifier.width(14.dp))
                Column(Modifier.padding(bottom = 22.dp)) {
                    Text(e.title, fontWeight = FontWeight.SemiBold,
                        fontSize = 15.sp)
                    Spacer(Modifier.height(2.dp))
                    Text(e.subtitle, color = scheme.onSurfaceVariant,
                        fontSize = 13.sp)
                }
            }
        }
    }
}
